doc/neps/nep-0042-new-dtypes.rst

```diff
@@ -0,0 +1,796 @@
+and some additional necessary information will thus be passed within a slots
+struct and identified by ``ssize_t`` integers::
+
+    static struct PyArrayMethodDef slots[] = {
+        {NPY_dt_method, method_implementation},
+        ...,
+        {0, NULL}
+    }
+
+    typedef struct{
+      PyTypeObject *typeobj;    /* type of python scalar */
+      int flexible;             /* Is the dtype flexible? */
+      int abstract;             /* Is the dtype abstract? */
+      int flags;                /* Currently only the "needs API" flag */
+      /* NULL terminated CastingImpl; is copied and references are stolen */
+      CastingImpl *castingimpls[];
+      PyType_Slot *slots;
+    } PyArrayDTypeMeta_Spec;
+
+    PyObject* PyArray_InitDTypeMetaFromSpec(
+            PyArray_DTypeMeta *user_dtype, PyArrayDTypeMeta_Spec *dtype_spec);
+
+all of this information will be copied during instantiation.
+
+The proposed method slots are (prepended with ``NPY_dt_``), these are
+detailed above and given here for summary:
+
+* ``is_native(self) -> {0, 1}``
+* ``ensure_native(self) -> dtype``
+* ``default_descr(self) -> dtype`` (return must be native and should normally be a singleton)
+* ``get_sort_function(self, NPY_SORTKIND sort_kind, void **out_sortfunction) -> success {-1, 0}``.
+  ``out_sortfunction`` may be ``NULL`` if the sort kind is unknown or not implemented.
+* ``setitem(self, char *item_ptr, PyObject *value) -> {-1, 0}``
+* ``getitem(self, char *item_ptr, PyObject (base_obj) -> object or NULL``
+* ``discover_descr_from_pyobject(cls, PyObject) -> dtype or NULL``
+* ``common_dtype(cls, other) -> DType, NotImplemented, or NULL``
+* ``common_instance(self, other) -> dtype or NULL``
+
+If not set, most slots are filled with slots which either error or defer automatically.
+Non-flexible dtypes do not have to implement:
+
+* ``discover_descr_from_pyobject`` (uses ``default_descr`` instead)
+* ``common_instance`` (uses ``default_descr`` instead)
+* ``ensure_native`` (uses ``default_descr`` instead)
+
+Which will be correct for most dtypes *which do not store metadata*.
+
+Other slots may be replaced by convenience versions, e.g. sorting methods
+can be defined by providing:
+
+* ``compare(self, char *item_ptr1, char *item_ptr2, int *res) -> {-1, 0}``
+  *TODO: We would like an error return, is this reasonable? (similar to old
+  python compare)*
+
+which uses generic sorting functionality.
+
+
+CastingImpl
+"""""""""""
+
+The external API for ``CastingImpl`` will be limited initially to defining:
+
+* ``cast_kind`` attribute, which can be one of the supported casting kinds.
+  This is the safest cast possible. For example casting between two NumPy
+  strings is of course "safe" in general, but may be "same kind" in a specific
+  instance if the second string is shorter. If neither type is flexible this
+  ``adjust_descriptors`` must use it. 
+* ``adjust_descriptors(dtypes_in[2], dtypes_out[2], casting) -> int {0, -1}``
```

Comment:
Note to self: After talking to Warren, I it may makes sense if we return `casting`, that way we could in theory add new casting more easily. It also pushes the natural place to the construct the casting error out of the `adjust_descriptors` function, where there is more context to construct an error message relevant to the user.

Reply:
Yes, I like that idea.